cpp/src/tsne/fft_tsne.cuh

```diff
@@ -0,0 +1,384 @@
+              const float post_learning_rate, const int max_iter,
+              const float min_grad_norm, const float pre_momentum,
+              const float post_momentum, const long long random_state,
+              const bool initialize_embeddings) {
+  auto d_alloc = handle.get_device_allocator();
+  auto stream = handle.get_stream();
+
+  // Get device properites
+  //---------------------------------------------------
+  const int mp_count = raft::getMultiProcessorCount();
+  const int dev_major_version = MLCommon::getDeviceCapability().first;
+  // These came from the CannyLab implementation, but I don't know how they were
+  // determined. TODO check/optimize.
+  const int integration_kernel_factor =
+    dev_major_version >= 6
+      ? 2
+      : dev_major_version == 5 ? 1 : dev_major_version == 3 ? 2 : 3;
+
+  constexpr int n_interpolation_points = 3;
+  constexpr int min_num_intervals = 50;
+  // The number of "charges" or s+2 sums i.e. number of kernel sums
+  constexpr int n_terms = 4;
+  int n_boxes_per_dim = min_num_intervals;
+
+  // FFTW is faster on numbers that can be written as 2^a 3^b 5^c 7^d 11^e 13^f
+  // where e+f is either 0 or 1, and the other exponents are arbitrary
+  int allowed_n_boxes_per_dim[20] = {25,  36,  50,  55,  60,  65,  70,
+                                     75,  80,  85,  90,  96,  100, 110,
+                                     120, 130, 140, 150, 175, 200};
+  if (n_boxes_per_dim < allowed_n_boxes_per_dim[19]) {
+    // Round up to nearest grid point
+    int chosen_i = 0;
+    while (allowed_n_boxes_per_dim[chosen_i] < n_boxes_per_dim) chosen_i++;
+    n_boxes_per_dim = allowed_n_boxes_per_dim[chosen_i];
+  }
+
+  int n_total_boxes = n_boxes_per_dim * n_boxes_per_dim;
+  int total_interpolation_points =
+    n_total_boxes * n_interpolation_points * n_interpolation_points;
+  int n_fft_coeffs_half = n_interpolation_points * n_boxes_per_dim;
+  int n_fft_coeffs = 2 * n_interpolation_points * n_boxes_per_dim;
+  int n_interpolation_points_1d = n_interpolation_points * n_boxes_per_dim;
+
+#define DB(type, name, size) \
+  MLCommon::device_buffer<type> name(d_alloc, stream, size)
+
+  DB(float, repulsive_forces_device, n * 2);
+  MLCommon::LinAlg::zero(repulsive_forces_device.data(),
+                         repulsive_forces_device.size(), stream);
+  DB(float, attractive_forces_device, n * 2);
+  DB(float, gains_device, n * 2);
+  auto gains_device_thrust = thrust::device_pointer_cast(gains_device.data());
+  thrust::fill(thrust::cuda::par.on(stream), gains_device_thrust,
+               gains_device_thrust + n * 2, 1.0f);
+  DB(float, old_forces_device, n * 2);
+  MLCommon::LinAlg::zero(old_forces_device.data(), old_forces_device.size(),
+                         stream);
+  DB(float, normalization_vec_device, n);
+  DB(int, point_box_idx_device, n);
+  DB(float, x_in_box_device, n);
+  DB(float, y_in_box_device, n);
+  DB(float, y_tilde_values, total_interpolation_points *n_terms);
```

Comment:
The formatter is insistent on the `a *b` spacing, which makes these look like pointer derefs. 😢